File: src/test_suite_oracle.py

from abc import ABC, abstractmethod
import graph

class TestSuiteOracle:

    @abstractmethod
    def run_tests(self, tests: list[int]) -> list[bool]:
        pass


class DirectDependenciesOracle(TestSuiteOracle):

    graph_generator: graph.GraphGenerator

    def __init__(self, n: int, graph_generator: graph.GraphGenerators) -> None:
        if graph_generator == graph.GraphGenerators.ErdosRenyi:
            self.graph_generator = graph.ErdosRenyiGenerator(n)
        elif graph_generator == graph.GraphGenerators.BarabasiAlbert:
            self.graph_generator = graph.BarabasiAlbertGenerator(n)
        else:
            raise ValueError("Not a valid graph generation algorithm")
        self.graph_generator.generate_edges()
        self.graph_generator.G.transitive_reduction()
        with open('./results/real_graph.gv', 'w') as fp:
            print(str(self.graph_generator.G), file=fp)

    def run_tests(self, tests: list[int]) -> list[bool]:
        results = []

        for i in range(len(tests)):
            deps = self.graph_generator.G.get_dependencies(tests[i])
            results.append(deps.intersection(set(tests[:i])) == deps)

        return results

File: src/test-suite-oracle.cc
#include "test-suite-oracle.h"
#include <cstdint>
#include <memory>

DirectDependenciesOracle::DirectDependenciesOracle(
    const std::vector<uint32_t> &nodes, const GraphGeneratorParams &params)
    : graph{nodes} {

  std::unique_ptr<GraphGenerator> graph_generator;
  if (params.generator_type == "erdos-renyi")
    graph_generator.reset(new ErdosRenyiGenerator{graph, params.probability});
  else if (params.generator_type == "barabasi-albert")
    graph_generator.reset(new BarabasiAlbertGenerator{graph});
  else if (params.generator_type == "out-degree")
    graph_generator.reset(
        new OutDegreeGenerator{graph, params.min_out, params.max_out});
  else {
    std::cerr << '"' << params.generator_type << '"'
              << " is not a valid method to generate graphs." << std::endl;
    exit(EXIT_FAILURE);
  }
  graph_generator->generate_edges();
}

DirectDependenciesOracle::DirectDependenciesOracle(const Graph &g)
    : graph{g}, test_suite_runs{0}, test_runs{0} {}

std::vector<bool>
DirectDependenciesOracle::run_tests(const std::vector<uint32_t> &tests) {
  std::vector<bool> results(tests.size(), false);

  ++test_suite_runs;

  for (uint32_t i = 0; i < tests.size(); ++i) {
    std::unordered_set<uint32_t> deps = graph.get_dependencies(tests[i]);

    for (const uint32_t dep : deps) {
      bool satisfied{false};
      for (uint32_t j = 0; j != i; ++j)
        if (tests[j] == dep)
          satisfied = true;

      if (!satisfied) {
        test_runs += i + 1;
        return results;
      }
    }

    results[i] = true;
  }

  test_runs += tests.size();

  return results;
}

std::vector<uint32_t> DirectDependenciesOracle::tests(void) const {
  std::vector<uint32_t> v;
  for (const auto &it : graph)
    v.push_back(it.first);
  return v;
}
